src/crimson/os/seastore/seastore.cc

```diff
@@ -1041,141 +1040,191 @@ SeaStore::tm_ret SeaStore::_do_transaction_step(
-      return _setattrs(ctx, get_onode(op->oid), std::move(to_set));
-    }
-    break;
-    case Transaction::OP_RMATTR:
-    {
-      std::string name = i.decode_string();
-      return _rmattr(ctx, get_onode(op->oid), name);
-    }
-    break;
-    case Transaction::OP_RMATTRS:
-    {
-      return _rmattrs(ctx, get_onode(op->oid));
-    }
-    break;
-    case Transaction::OP_MKCOLL:
-    {
-      coll_t cid = i.get_cid(op->cid);
-      return _create_collection(ctx, cid, op->split_bits);
-    }
-    break;
+  if (op->op == Transaction::OP_NOP)
+    return tm_iertr::now();
+
+  switch (op->op) {
     case Transaction::OP_RMCOLL:
     {
       coll_t cid = i.get_cid(op->cid);
       return _remove_collection(ctx, cid);
     }
     break;
-    case Transaction::OP_OMAP_SETKEYS:
-    {
-      std::map<std::string, ceph::bufferlist> aset;
-      i.decode_attrset(aset);
-      return _omap_set_values(ctx, get_onode(op->oid), std::move(aset));
-    }
-    break;
-    case Transaction::OP_OMAP_SETHEADER:
-    {
-      ceph::bufferlist bl;
-      i.decode_bl(bl);
-      return _omap_set_header(ctx, get_onode(op->oid), std::move(bl));
-    }
-    break;
-    case Transaction::OP_OMAP_RMKEYS:
-    {
-      omap_keys_t keys;
-      i.decode_keyset(keys);
-      return _omap_rmkeys(ctx, get_onode(op->oid), std::move(keys));
-    }
-    break;
-    case Transaction::OP_OMAP_RMKEYRANGE:
+    case Transaction::OP_MKCOLL:
     {
-      string first, last;
-      first = i.decode_string();
-      last = i.decode_string();
-      return _omap_rmkeyrange(
-	ctx, get_onode(op->oid),
-	std::move(first), std::move(last));
+      coll_t cid = i.get_cid(op->cid);
+      return _create_collection(ctx, cid, op->split_bits);
     }
     break;
-    case Transaction::OP_OMAP_CLEAR:
-    {
-      return _omap_clear(ctx, get_onode(op->oid));
-    }
     case Transaction::OP_COLL_HINT:
     {
       ceph::bufferlist hint;
       i.decode_bl(hint);
       return tm_iertr::now();
     }
-    case Transaction::OP_ZERO:
-    {
-      objaddr_t off = op->off;
-      extent_len_t len = op->len;
-      return _zero(ctx, get_onode(op->oid), off, len);
+  }
+
+  using onode_iertr = OnodeManager::get_onode_iertr::extend<
+    crimson::ct_error::value_too_large>;
+  auto fut = onode_iertr::make_ready_future<OnodeRef>(OnodeRef());
+  bool create = false;
+  if (op->op == Transaction::OP_TOUCH ||
+      op->op == Transaction::OP_CREATE ||
+      op->op == Transaction::OP_WRITE ||
+      op->op == Transaction::OP_ZERO) {
+    create = true;
+  }
+  if (!onodes[op->oid]) {
+    if (!create) {
+      fut = onode_manager->get_onode(*ctx.transaction, i.get_oid(op->oid));
+    } else {
+      fut = onode_manager->get_or_create_onode(
+        *ctx.transaction, i.get_oid(op->oid));
     }
-    default:
-      ERROR("bad op {}", static_cast<unsigned>(op->op));
+  }
+  return fut.si_then([&, op, this](auto&& get_onode) -> tm_ret {
+    OnodeRef &o = onodes[op->oid];
+    if (get_onode) {
```

Comment:
Seems `get_onode` should always have value or `enoent` raised, so just `assert(get_onode)` here, right?

Reply:
enoent will be handled in handle_error_interruptible.

Reply:
So `get_onode` here must be available, so just `assert(get_onode)`.